src/org/sosy_lab/java_smt/example/NQueens.java

```diff
@@ -0,0 +1,208 @@
+
+                logger.logUserException(Level.INFO, e, "Solver " + solver + " is not available.");
+
+            } catch (UnsupportedOperationException e) {
+                logger.logUserException(Level.INFO, e, e.getMessage());
+            }
+        }
+    }
+}
+abstract class NQueensSolver {
+
+    private final SolverContext context;
+    final BooleanFormulaManager bmgr;
+
+    NQueensSolver(SolverContext pContext) {
+        context = pContext;
+        bmgr = context.getFormulaManager().getBooleanFormulaManager();
+    }
+    abstract BooleanFormula[][] getSymbols();
+    abstract List<BooleanFormula> getRules(BooleanFormula[][] symbols, SolverContext context);
+    abstract Boolean getValue(BooleanFormula[][] symbols, Model model, int row, int col);
+
+    /**
+     * Solves the N-Queens problem for the given board size and returns a possible solution.
+     * Returns <code>Null</code> if no solution exists.
+     */
+    @Nullable
+    public Boolean[][] solve(int n) throws InterruptedException, SolverException {
+        BooleanFormula[][] symbols = getSymbols();
+        List<BooleanFormula> rules = getRules(symbols, context);
+        // solve N-Queens
+        try (ProverEnvironment prover = context.newProverEnvironment(ProverOptions.GENERATE_MODELS)) {
+            prover.push(bmgr.and(rules));
+            boolean isUnsolvable = prover.isUnsat();
+            if (isUnsolvable) {
+                return null;
+            }
+
+            // get model and convert it
+            Boolean[][] solution = new Boolean[n][n] ;
+            try (Model model = prover.getModel()) {
+                for (int row = 0; row < n; row++) {
+                    for (int col = 0; col < n; col++) {
+                        solution[row][col] = getValue(symbols, model, row,col);
+                    }
+                }
+                return solution;
+            }
+        }
+    }
+}
+class NQueen extends NQueensSolver {
+    private final int n;
+
+    public NQueen(SolverContext context, int n) {
+        super(context);
+        this.n = n;
+    }
+
+    /**
+     * prepare symbols: one symbol for each of the N*N cells.
+     */
+    @Override
+    BooleanFormula[][] getSymbols() {
+        final BooleanFormula[][] symbols = new BooleanFormula[n][n];
+        for (int row = 0; row < n; row++) {
+            for (int col = 0; col < n; col++) {
+                symbols[row][col] = bmgr.makeVariable("q_" + row + "_" + col);
+            }
+        }
+        return symbols;
+    }
+
+    /*
+     * getRules is the method used to add constraints that ensure that no two queens are in the same
+     * row, column, or diagonal.
+     */
+    @Override
+    List<BooleanFormula> getRules(BooleanFormula[][] symbols, SolverContext context) {
+        List<BooleanFormula> rules = new ArrayList<>();
+
+        // Add constraints to ensure that only one queen is placed in each row
```

Comment:
The following lines produce the following constraint for each row 'r': `q_r_0  OR  q_r_1  OR  q_r_2  OR  q_r_3`.
This contraint does require that there is at least one queen,
but it does not forbit that there are multiple queens.

A solver can simply assume that every cell has a queen,
...which would be allowed by the constraint, but invalid for NQueens problem.
The constraint needs to be improved.